File: src/util/demo/IconExtractorDemo.java
package util.demo;

import java.io.File;

import javafx.application.Application;
import javafx.collections.ObservableList;
import javafx.scene.Scene;
import javafx.scene.control.ListCell;
import javafx.scene.control.ListView;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.layout.Priority;
import javafx.scene.layout.VBox;
import javafx.stage.Stage;
import util.graphics.IconExtractor;

import static javafx.collections.FXCollections.observableArrayList;

public class IconExtractorDemo extends Application {

	public static void main(String[] args) {
		launch(args);
	}

	private final ListView<String> list = new ListView<>();
	private final ObservableList<String> data = observableArrayList(
		// "C:\\CCleaner\\CCleaner.exe", // an example to test out .exe (windows executable) file
		// "C:\\CCleaner\\CCleaner.lnk", // an example to test out .lnk (windows shortcut) file
		"a.msg", "a1.msg", "b.txt", "c.pdf",
		"d.html", "e.png", "f.zip",
		"g.docx", "h.xlsx", "i.pptx"
	);

	@Override
	public void start(Stage stage) {
		VBox box = new VBox();
		Scene scene = new Scene(box, 200, 200);
		stage.setScene(scene);
		stage.setTitle("ListViewSample");
		box.getChildren().addAll(list);
		VBox.setVgrow(list, Priority.ALWAYS);

		list.setItems(data);
		list.setCellFactory(list1 -> new AttachmentListCell());

		stage.show();
	}

	private static class AttachmentListCell extends ListCell<String> {
		@Override
		public void updateItem(String item, boolean empty) {
			super.updateItem(item, empty);
			if (empty) {
				setGraphic(null);
				setText(null);
			} else {
				Image fxImage = IconExtractor.getFileIcon(new File(item));
				ImageView imageView = new ImageView(fxImage);
				setGraphic(imageView);
				setText(item);
			}
		}
	}

}

File: src/demo/main/sp/it/demo/IconExtractorDemo.kt
package sp.it.demo

import java.io.File
import javafx.application.Application
import javafx.geometry.Pos.CENTER
import javafx.scene.Scene
import javafx.scene.control.ListCell
import javafx.scene.image.ImageView
import javafx.scene.input.DragEvent.DRAG_DROPPED
import javafx.scene.input.DragEvent.DRAG_OVER
import javafx.scene.input.TransferMode
import javafx.scene.layout.Priority.ALWAYS
import javafx.stage.Stage
import javafx.util.Callback
import sp.it.util.ui.IconExtractor.getFileIcon
import sp.it.util.ui.label
import sp.it.util.ui.lay
import sp.it.util.ui.listView
import sp.it.util.ui.vBox

/** Showcases the icon extraction for various file types and files */
class IconExtractorDemo: Application() {

   private val list = listView<String> {
      items += listOf(
         "a.msg", "a1.msg", "b.txt", "c.pdf", "d.html", "e.png", "f.zip", "g.docx", "h.xlsx", "i.pptx"
      )
      cellFactory = Callback {
         object: ListCell<String>() {
            override fun updateItem(item: String?, empty: Boolean) {
               super.updateItem(item, empty)
               if (empty || item==null) {
                  graphic = null
                  text = null
               } else {
                  graphic = ImageView(getFileIcon(File(item)))
                  text = item
               }
            }
         }
      }
   }

   override fun start(stage: Stage) {
      val box = vBox {
         lay += label("Drag & drop files (such as .exe) to display icon") { alignment = CENTER }
         lay(ALWAYS) += list

         addEventHandler(DRAG_OVER) {
            if (it.dragboard.hasFiles()) {
               it.acceptTransferModes(*TransferMode.ANY)
               it.consume()
            }
         }
         addEventHandler(DRAG_DROPPED) {
            if (it.dragboard.hasFiles()) {
               list.items += it.dragboard.files.map { it.absolutePath }
               it.isDropCompleted = true
               it.consume()
            }
         }
      }

      stage.scene = Scene(box, 200.0, 200.0)
      stage.title = "ListViewSample"
      stage.show()
   }
}

fun main(args: Array<String>) {
   Application.launch(IconExtractorDemo::class.java, *args)
}